### ISSUE TITLE
Notes

### ISSUE THREAD

**Issue by PedroRegisPOAR** [ES-Nix/locale#1]
## Abstract

There are 4 packages in nixpkgs involved, at least.

```bash
nix build nixpkgs#glibcLocales
```

```bash
nix build nixpkgs#glibc.bin
```

```bash
nix build nixpkgs#locale
```

```bash
nix build nixpkgs#glibc
```

### Some notes about locale/locale-archive

- domenkozar opened this issue on Mar 18, 2015 · 15 comments: https://github.com/NixOS/nixpkgs/issues/6878
- trofi opened this issue on Jan 11, 2017 · 10 comments: https://github.com/NixOS/nixpkgs/issues/21820
- https://github.com/NixOS/nixpkgs/issues/60414
- https://github.com/nix-community/home-manager/issues/2333
- https://stackoverflow.com/questions/5909404/sort-not-sorting-as-expected-space-and-locale/5909526#5909526
- https://stackoverflow.com/questions/23660247/treatment-of-spaces-in-sort-command-difference-between-lc-collate-c-and-lc-coll?rq=1
- https://mpickering.github.io/posts/2018-09-19-nix-artefacts.html "LOCALE_ARCHIVE=${glibcLocales}/lib/locale/locale-archive"

Old, but really great: https://github.com/NixOS/nix/issues/599#issuecomment-153885553

TODO: take a look in this, lots of troubleshoot commands:
https://github.com/NixOS/nix/issues/599

### What is the locale-archive

> `locale-archive` is a memory-mapped file which is generated by `locale-gen(8)` invoking `localedef(1)`. Memory-mapped means that once it is created and called by a program it is only loaded once into memory.
> https://unix.stackexchange.com/a/331706

[Difference between locale-archive and Machine Object files in /usr/share/locale/<locale_dir>/LC_MESSAGES/ directory?](https://unix.stackexchange.com/questions/330345/difference-between-locale-archive-and-machine-object-files-in-usr-share-locale)

TODO: add real updated values and sha256sum
> In the above case deleting all locales except the en_* ones my locale-archive went from 102MB down to 3.4MB
> https://unix.stackexchange.com/a/498523

> `find . ! -name 'file.txt' -type f -exec rm -f {} +` 
> From: https://unix.stackexchange.com/a/153863

#### glibc

```nix
glibc = super.glibc.overrideAttrs (_: {
      # Warning: MASSIVE rebuild since you'll break ABI
      version = "2.26";
    });
```
From: https://gurkan.in/wiki/nix.html#override-example-optional-args

TODO: convert this to a flake [nix-shell cannot change locale warning](https://stackoverflow.com/a/63123144)

Some troubleshoot commands:
-  `strace -e file locale`
-  `ls /usr/share/i18n/charmaps/`
-  `gzip -dk UTF-8.gz`
From: [Locale issue after upgrading to Ubuntu 16.10 from a clean installation of Ubuntu 16.04](https://github.com/microsoft/WSL/issues/1232)

Saving all this for now:
- [setlocale() fails for anything other than "C" (missing locale-archive?)](https://github.com/NixOS/nixpkgs/issues/39328)
- [glibc 2.27 breaks locale support](https://github.com/NixOS/nixpkgs/issues/38991)
- [Make NixOS provide version-specific LOCALE_ARCHIVE environment variables](https://gist.github.com/peti/2c818d6cb49b0b0f2fd7c300f8386bc3)
- [glibc 2.27 breaks locale support](https://github.com/NixOS/nixpkgs/issues/38991)
- [Change the order of locale archive search path](https://github.com/NixOS/nixpkgs/issues/85823)
- [A function to build Glibc's locale-archive separately from Glibc.](https://git.stoo.org/infra/nixpkgs/commit/40f01daa819a05b956cdca0b0abdfa9f07c9e327)
- https://github.com/NixOS/nix/issues/3579#issuecomment-626118265
- https://askubuntu.com/a/1028473
- https://github.com/NixOS/nixpkgs/issues/170897

>  I think the proper fix would be to include en_US.utf-8 in our glibc version by default (can be done via overrideAttrs) and rebuild python/bash/perl against that.
[Build a yocto rootfs inside nix](https://discourse.nixos.org/t/build-a-yocto-rootfs-inside-nix/2643/22)

  >  Force install locale from "glibcLocales" since there are collisions
  > extraBuildCommands = ''
  >   ln -sf ${glibcLocales}/lib/locale/locale-archive $out/usr/lib/locale
  > '';
[stammw/yocto.nix](https://gist.github.com/stammw/d55958fc7ac6185e3446ccf637e53683#file-yocto-nix-L54)

> After evaluating this issue, there are no plans to address it further or fix it in an upcoming release.  Therefore, it is being closed.  If plans change such that this issue will be fixed in an upcoming release, then the bug can be reopened.
[Unable to invoke localedef to UTF-8: cannot read character map directory `/usr/share/i18n/charmaps'](https://bugzilla.redhat.com/show_bug.cgi?id=1677927#c10)

TODO: https://unix.stackexchange.com/questions/187402/nix-package-manager-perl-warning-setting-locale-failed/243189#243189

Maybe useful:
https://github.com/davidtwco/veritas/blob/6f2c676a76ef2885c9102aeaea874c361dbcaf61/home/profiles/common.nix#L197-L198

TODO: document from where it came from, the python PEPs about it
```bash
nix run nixpkgs#python3 -- <<<'print("ℙƴ☂ℌøἤ")'
```
- https://peps.python.org/pep-0538/
- https://peps.python.org/pep-0540/

> At the moment, setting "LANG=C" on a Linux system *fundamentally breaks Python 3*, and that's not OK.
> ENV LANG C.UTF-8
https://github.com/docker-library/python/blob/56cea612ab370f3d05b29e97466d418a0f07e463/3.10/slim-bullseye/Dockerfile#L12-L14

https://click.palletsprojects.com/en/5.x/python3/#python-3-surrogate-handling

**Comment by PedroRegisPOAR** [comment]
### Abstract


> To explain, the default glibc package doesn't contain locale data – that's separate and --pure isn't supposed to see OS data.
Source: https://github.com/NixOS/nixpkgs/issues/32848#issuecomment-352996633


> Adding `glibcLocales` to the shell indeed fixes the issue. Though this raises the question
as to whether that dependency needs to be explicitly added to any package that depends
on glibc and does some kind of text processing. Locales are a non-optional part of the C
standard and while it’s great to be able to drop the heavyweight dependency where you
know it’s irrelevant, it should not be absent in the default context.
From: 
- https://github.com/NixOS/nixpkgs/issues/60414#issuecomment-679045623


#### MWE

##### MWE 1:

```bash
LC_ALL=en_GB.utf8 date '+%c'
LC_ALL=en_US.utf8 date '+%c' 
```

```bash
nix \
run \
nixpkgs#python39 \
-- \
-c \
'
import locale
locale.setlocale(locale.LC_ALL, "pt_BR.utf8")
print(locale.currency(12345.67, grouping=True, symbol=True))
'
```


#### copy/pastes


##### Code 1:
```bash
# 
# nix flake metadata nixpkgs --json | jq -r .url
nix \
store \
ls \
--store https://cache.nixos.org/ \
--long \
--recursive \
"$(nix eval --raw github:NixOS/nixpkgs/9eb60f25aff0d2218c848dd4574a0ab5e296cabe#glibcLocales)"
```

##### Code 2:
```bash
# 
# nix flake metadata nixpkgs --json | jq -r .url
nix \
store \
ls \
--store https://cache.nixos.org/ \
--long \
--recursive \
"$(nix eval --raw github:NixOS/nixpkgs/9eb60f25aff0d2218c848dd4574a0ab5e296cabe#glibc)"
```


##### Code 3:
```bash
# 
# nix flake metadata nixpkgs --json | jq -r .url
nix \
store \
ls \
--store https://cache.nixos.org/ \
--long \
--recursive \
"$(nix eval --raw github:NixOS/nixpkgs/9eb60f25aff0d2218c848dd4574a0ab5e296cabe#locale)"
```

##### Code 4:
```bash
# 
# nix flake metadata nixpkgs --json | jq -r .url
nix \
store \
ls \
--store https://cache.nixos.org/ \
--long \
--recursive \
"$(nix eval --raw github:NixOS/nixpkgs/9eb60f25aff0d2218c848dd4574a0ab5e296cabe#glibc.bin)"
```


```bash
# nix flake metadata github:NixOS/nixpkgs/release-22.05 --json
command -v jq >/dev/null || nix profile install github:NixOS/nixpkgs/4aceab3cadf9fef6f70b9f6a9df964218650db0a#jq \
&& nix \
build \
--impure \
--expr \
'(
  with builtins.getFlake "nixpkgs";
  with legacyPackages.${builtins.currentSystem};
  (
    glibcLocales.override {
        allLocales = false;
        locales = [ 
                           "en_GB.UTF-8/UTF-8" 
                           "ru_RU.UTF-8/UTF-8" 
                           "en_US.UTF-8/UTF-8" 
                           "pt_BR.UTF-8/UTF-8"
                           "ja_JP.UTF-8/UTF-8"
                           "en_IE.UTF-8/UTF-8"
                     ];
      }
  )
)'
```
Refs.:
- https://github.com/NixOS/nixpkgs/pull/176253#issuecomment-1152892550
- https://unix.stackexchange.com/a/62317



```bash
LOCALE_ARCHIVE=result/lib/locale/locale-archive \
&& LC_ALL=pt_BT.UTF-8 \
&& nix \
run \
nixpkgs#python39 \
-- \
-c \
'
import locale

locale.setlocale(locale.LC_ALL, "pt_BR.utf8")
print(locale.currency(12345.67, grouping=True, symbol=True))

locale.setlocale(locale.LC_ALL, "en_US.utf8")
print(locale.currency(12345.67, grouping=True, symbol=True))

locale.setlocale(locale.LC_ALL, "ru_RU.utf8")
print(locale.currency(12345.67, grouping=True, symbol=True))

locale.setlocale(locale.LC_ALL, "ja_JP.utf8")
print(locale.currency(12345.67, grouping=True, symbol=True))

locale.setlocale(locale.LC_ALL, "en_IE.utf8")
print(locale.currency(12345.67, grouping=True, symbol=True))
'
```

```bash
{ cat <<WRAP >> foo.c
#include <stdio.h>
#include <locale.h>

int main()
{
    char *locale = setlocale(LC_ALL, "");
    printf("\n locale =%s\n", locale);
    printf("test\n \u263a\u263b Hello from C\n");

    return 0;
}
WRAP
} && gcc foo.c \
&& ./a.out
```

```bash
rm -f a.out foo.c
```
Refs.:
- https://stackoverflow.com/a/58071375


TODO: 
```bash
nix run nixpkgs#gcc -- -xc -E -v /dev/null
```
+
```bash
printf '#include <locale.h>\nLC_COLLATE\n' | gcc -E -x c - | tail -n 1
```
Refs.:
- https://unix.stackexchange.com/a/449236

```bash
nix run nixpkgs#python39 -- -c "assert '\N{snake}' == '🐍'"
```

TODO:
```bash
printf %b\\n \\u04{51,{3,4}{{0..9},{a..f}}}|sort|sed 's/./\u&&/'|tr -d \\n
```
Refs.:
- https://codegolf.stackexchange.com/a/127708

#### cowsay

```bash
nix \
shell \
--ignore-environment \
--impure \
--expr \
'
  (
    let
      nixpkgs = (builtins.getFlake "github:NixOS/nixpkgs/0938d73bb143f4ae037143572f11f4338c7b2d1c");
      pkgs = import nixpkgs { };    
    in
      with pkgs; [
        cowsay
      ]
    )
' \
--command cowsay "Hello"

nix \
shell \
--ignore-environment \
--impure \
--expr \
'
  (
    let
      nixpkgs = (builtins.getFlake "github:NixOS/nixpkgs/0938d73bb143f4ae037143572f11f4338c7b2d1c");
      pkgs = import nixpkgs { };    
    in
      with pkgs; [
        (
          glibcLocales.override {
              allLocales = false;
              locales = [
                          "en_US.UTF-8/UTF-8" 
                          "pt_BR.UTF-8/UTF-8"
                        ];
            }
        )
        cowsay
      ]
    )
' \
--command cowsay "Hello"
```
Refs.:
- [Matthew Croughan - What Nix Can Do (Docker Can't) - SCaLE 20x](https://www.youtube.com/embed/6Le0IbPRzOE?start=4388&end=4402&version=3), start=4388&end=4402


#### Others

```bash
LOCALE_ARCHIVE=result/lib/locale/locale-archive


LC_ALL=pt_BR.UTF-8 date '+%c'
LC_ALL=en_US.UTF-8 date '+%c'
LC_ALL=ru_RU.UTF-8 date '+%c'
LC_ALL=ja_JP.UTF-8 date '+%c'
LC_ALL=en_IE.UTF-8 date '+%c'
```



```bash
LOCALE_ARCHIVE=result/lib/locale/locale-archive


LC_ALL=en_GB.UTF-8
nix run nixpkgs#uutils-coreutils -- date '+%c'


LC_ALL=en_US.UTF-8
nix run nixpkgs#uutils-coreutils -- date '+%c'
```

```bash
LOCALE_ARCHIVE=result/lib/locale/locale-archive


LC_ALL=en_GB.UTF-8
nix run nixpkgs#busybox -- date '+%c'


LC_ALL=en_US.UTF-8
nix run nixpkgs#busybox -- date '+%c'
```


```bash
nix \
run \
nixpkgs#python39 -- \
-c \
"
v=32
while v:print('Ёё'*(v==26),end='%c%c'%(1072-v,1104-v));v-=1
"
```


```bash
export LC_ALL=en_US.utf8
nix run nixpkgs#python39 -- -c '
import locale
defaultlocale = locale.getdefaultlocale()
locale.setlocale(locale.LC_ALL, defaultlocale[0] + "." + defaultlocale[1])
print(locale.currency(12345.67, grouping=True, symbol=True))
'

export LC_ALL=pt_BR.utf8
nix run nixpkgs#python39 -- -c '
import locale
defaultlocale = locale.getdefaultlocale()
locale.setlocale(locale.LC_ALL, defaultlocale[0] + "." + defaultlocale[1])
print(locale.currency(12345.67, grouping=True, symbol=True))
'
```
Refs.:
- https://docs.python.org/3/library/locale.html#locale.getdefaultlocale



```bash
perl -MEncode=decode -E 'while(<>){ chomp; say length decode("UTF-8", $_) }' <<<'文字化け'
```
Refs.:
- https://stackoverflow.com/a/6192088

TODO: 
- how is this locale is computed, system locale or internal to the Faker python library?
- How about the python PEP and packages tzdata and zoneinfo https://docs.python.org/3/library/zoneinfo.html https://pypi.org/project/tzdata/
- [Live de Python #80 - Melhorando testes de unidade com Faker, Factory boy e Hypothesis](https://www.youtube.com/watch?v=HuZ2Keoc9Hs&t=875s), t=875s
- Java: https://stackoverflow.com/a/62186308
- https://bugs.python.org/issue19846
- https://bugs.python.org/issue28180

```bash
Faker("cellphone_number", locale="pt-BR")
```


TODO:
- https://github.com/systemd/systemd/blob/d9c6956ee4c10b76b4ed4ae32550e15a300172f3/mkosi.build#L27-L48 and  `systemd-escape -u 'Hall\xc3\xb6chen\x2c\x20Meister'`
- https://stackoverflow.com/a/43989185
- https://pythonmana.com/2022/01/202201010619318735.html
- https://stackoverflow.com/questions/63019217/write-unicode-text-on-image-in-python-pillow
- https://stackoverflow.com/questions/40302497/erratic-localization-behavior-of-gnu-gettext-compared-to-strftime-or-printf
- https://stackoverflow.com/a/31970670
- https://stackoverflow.com/q/63019217
- https://stackoverflow.com/questions/64790054/how-to-specify-localized-forms-of-cjk-scripts-in-pythons-pil-imagefont
- https://unix.stackexchange.com/questions/101844/printing-utf-8-including-chinese-text
- https://unix.stackexchange.com/a/233158

**Comment by PedroRegisPOAR** [comment]
TODO: 


> At 31:50 you talk about environment variables. However there are some mistakes worth correcting for future viewers. First, although the environment variables are stored in the process' memory, it is stored as zero-terminated strings and not as one big string separated by new-line characters. It is also is not stored on the heap, nor is there a global variable in the data section pointing to it. The environment is actually stored entirely on the stack and is a part of the initial process stack that is set up before the program starts running. The first value on the stack is the argument count followed by an array of the addresses of the different arguments, then address 0 marking the end of the argument array. Right after that there is a second array of addresses which each point to a zero-terminated string which would be the environment variables, this array is also terminated by having address 0 at the end. There is actually a third array of auxiliary vectors but after that there is an unspecified amount of bytes before the information block starts. It's generally inside this block the command line arguments and environment variables are stored, as in the actual string values. You can confirm this by dumping the stack of pretty much any program and you typically find all the environment variables at the very end (highest memory address). If you are on Linux you can do this by first reading the  '/proc/<pid>/maps' file for any process, just replace <pid> with that process' PID. This file contains the ranges of memory mapped to the process and what they are mapped to. Near the bottom you'll see one line with the range mapped to [stack]. Take note of the start address and calculate how big it is in bytes. Then run 'sudo xxd -s <start-address> -l <size-in-bytes> /dev/<pid>/mem', example 'sudo xxd -s 0x7fff182bd000 -l 0x22000 /dev/14950/mem'. And the environment variables should get printed out together with their hex values and address location.

To illustrate this further I've written a small c program that prints all the environment variables using the argv array pointer. As you can see the environment variable pointers are stored pretty much right after argv.

```c
#include <stdio.h>
int main(int argc, char **argv)
{
 for (int i = argc + 2; argv[i] != NULL; i++)
 {
  printf("%s\n", argv[i]);
 }
 return 0;
}
```

You can of course make it less stupid by using the full version of main which includes a pointer to the first element in the environment pointer array.

```c
#include <stdio.h>
int main(int argc, char **argv, char **envp)
{
 for (int i = 0; envp[i] != NULL; i++)
 {
  printf("%s\n", envp[i]);
 }
 return 0;
}
```
This is all defined as a part of the ABI (application binary interface) for both the x86 and x86_64 architecture, so 32 and 64 bit desktop computers.

tl;dr: The environment is not a single long string separated by new-line characters. The environment variables and the pointers to them are both stored on the stack or just before it.
https://www.youtube.com/watch?v=xHu7qI1gDPA&lc=UgwvbQ7HZFUEZ2EGQ7V4AaABAg

Maybe related: https://serverfault.com/a/792136


Related: [Dangerous Code Hidden in Plain Sight for 12 years](https://youtu.be/eTcVLqKpZJc?t=219)

**Comment by PedroRegisPOAR** [comment]
### Fonts, yes they are related too

Yeah, here i am trying to summarize all this mess.


> FC_LANG is used to specify the default language as the weak binding in the query. if this isn't set, the default language will be determined from current locale. 
https://www.freedesktop.org/software/fontconfig/fontconfig-user.html

```bash
export FONTCONFIG_PATH=/etc/fonts
```
Refs.:
- https://askubuntu.com/a/708541


TODOs:
- https://stackoverflow.com/questions/39507534/statically-linking-fontconfig-disable-fonts-conf-file
- https://askubuntu.com/questions/202389/per-user-fonts-conf-file
- https://en.wikipedia.org/wiki/Fontconfig
- https://en.wikipedia.org/wiki/X_logical_font_description
- https://wiki.archlinux.org/title/XDG_Base_Directory
- [How do I avoid mojibake in NixOS?](https://discourse.nixos.org/t/how-do-i-avoid-mojibake-in-nixos/8261)

```bash
man fonts-conf
```

```bash
ls -al $(nix build --no-link --print-build-logs --print-out-paths github:NixOS/nixpkgs/0938d73bb143f4ae037143572f11f4338c7b2d1c#xorg.fontalias)/share/fonts/X11
```

#### Some really good videos

- [Characters, Symbols and the Unicode Miracle - Computerphile](https://www.youtube.com/watch?v=MijmeoH9LT4)
- [Code Pages, Character Encoding, Unicode, UTF-8 and the BOM - Computer Stuff They Didn't Teach You #2](https://www.youtube.com/watch?v=jeIBNn5Y5fI)
- [What's a Carriage and Who's Feeding it Lines? CRLF - Computer Stuff They Didn't Teach You #1](https://www.youtube.com/watch?v=TtiBhktB4Qg)
- [Internationalis(z)ing Code - Computerphile](https://www.youtube.com/watch?v=0j74jcxSunY)
- [The Problem with Time & Timezones - Computerphile](https://www.youtube.com/watch?v=-5wpm-gesOY)


#### Minimal working example (on my NixOS machine with zsh)

```bash
echo '\u2603'

LC_CTYPE=C echo '\u2603'
```
Refs.:
- https://github.com/ohmyzsh/ohmyzsh/issues/4065#issuecomment-115497516


Errors out:
```bash
zsh: character not in range
```

> Also, `UTF-8` is not a valid POSIX locale. It may work on some systems, but Arch Linux might not like it. `en_US.UTF-8` is valid. Try putting that at the beginning of the line, and using `LC_ALL` instead of `LC_CTYPE`.
https://github.com/ohmyzsh/ohmyzsh/issues/4065#issuecomment-129913471

> We ended up just giving up on trying to fix broken locales, thanks for your contribution and your patience.
https://github.com/ohmyzsh/ohmyzsh/pull/4696#issuecomment-537132617

### Must read

Main ones:
- https://medium.com/@wanxiao1994/some-trivial-knowledge-about-unicode-4f70750af0ba
- https://esham.io/2014/06/unicode-flags
- https://esham.io/2015/04/emoji-flags-redux
- https://superuser.com/a/1266865
- https://shkspr.mobi/blog/2019/06/quirks-and-limitations-of-emoji-flags/
- https://qz.com/2148267/the-reason-why-unicode-wont-add-new-flag-emoji
- https://superuser.com/questions/1676773/how-do-i-preview-a-string-containing-unsupported-unicode-characters/1676816#1676816
- https://unix.stackexchange.com/questions/535858/test-for-terminal-emoji-support#comment993467_535858
- https://github.com/ohmyzsh/ohmyzsh/issues/4065#issuecomment-174783649
- https://github.com/ohmyzsh/ohmyzsh/issues/4065#issuecomment-129913471
- https://github.com/ohmyzsh/ohmyzsh/issues/4065#issuecomment-139384892
- https://github.com/ohmyzsh/ohmyzsh/issues/4065#issuecomment-174783649
- https://unix.stackexchange.com/a/285002
- https://github.com/NixOS/nixpkgs/issues/86713#issuecomment-792133582
- https://github.com/NixOS/nixpkgs/issues/86601#issuecomment-757959059
- https://spaceship-prompt.sh/faq/
- https://superuser.com/questions/1545566/is-there-any-way-to-replace-segoe-ui-emoji-that-comes-with-windows-with-a-font-t

Tables:
- https://unicode.org/Public/emoji/
- https://github.com/posoo/emojify-zsh/blob/53e13ef234b713e9d4839c5b4e0bdb375fed6fe7/emojify#L28-L2793
- https://github.com/DiegoFleitas/DeviantBot/wiki/Emojis
- https://getemoji.com/
- https://apps.timwhitlock.info/emoji/tables/iso3166
- https://emojipedia.org/emoji-flag-sequence/
- https://en.wikipedia.org/wiki/Regional_indicator_symbol
- https://tug.ctan.org/info/symbols/comprehensive/symbols-a4.pdf


Fonts:
- https://stackoverflow.com/a/40896083
- https://unicode.org/resources/fonts.html
- https://blog.emojipedia.org/exploring-googles-new-black-and-blobby-emoji-font/
- https://fonts.google.com/noto/use#how-are-noto-fonts-organized
- https://www.babelstone.co.uk/Fonts/Flags.html
- https://www.fontspace.com/unicode/char/200C-zero-width-non-joiner
- https://github.com/adobe-fonts/emojione-color
- https://stackoverflow.com/a/72731942
- https://superuser.com/a/1266892
- https://nolanlawson.com/2022/04/08/the-struggle-of-using-native-emoji-on-the-web/
- https://superuser.com/a/1671164
- TODO: https://askubuntu.com/a/740649

More python focused:
- https://docs.python.org/3/library/unicodedata.html
- https://docs.python.org/3/howto/unicode.html
- https://www.geeksforgeeks.org/unicodedata-unicode-database-python/
- https://www.geeksforgeeks.org/unicode_literals-in-python/
- https://www.geeksforgeeks.org/python-convert-string-to-unicode-characters/
- https://stackoverflow.com/a/44457367
- https://stackoverflow.com/a/58200269
- https://github.com/cvzi/flag
- https://www.johndcook.com/blog/2022/10/02/flags-unicode/
- https://www.fluentpython.com/extra/multi-character-emojis/
- https://note.nkmk.me/en/python-chr-ord-unicode-code-point/
- https://davidamos.dev/why-cant-you-reverse-a-flag-emoji/
- https://gist.github.com/treyhunner/8261108b7d333132b544807ee3ffd55a

Even Julia:
- https://discourse.julialang.org/t/recommendation-for-a-font-with-good-unicode-support/39885

Of course LaTeX:
- https://tex.stackexchange.com/questions/352623/unicode-char-u200b
- https://pt.overleaf.com/learn/latex/Questions/Inserting_emojis_in_LaTeX_documents_on_Overleaf


Linux:
- https://forum.manjaro.org/t/howto-enable-emoji-fonts/36695/6
- https://chrpaul.de/2019/07/Enable-colour-emoji-support-on-Manjaro-Linux.html
- https://askubuntu.com/questions/1130880/color-emojis-working-partially-on-18-04
- https://github.com/maoschanz/emoji-selector-for-gnome/issues/86
- https://pkgs.alpinelinux.org/package/edge/community/x86/font-noto-emoji
- https://briancaffey.github.io/2017/11/24/how-to-get-color-emoji-in-arch-linux.html/
- https://forum.manjaro.org/t/howto-enable-emoji-fonts/36695/4
- https://opensource.com/article/21/2/emoji-linux
- https://victor.kropp.name/blog/emoji-on-linux/
- https://mobile.twitter.com/XTaran/status/1536751836425756675


### Troubleshooting


```bash
fc-cache -fv
```

```bash
echo "\ue0b0 \ue0a0 \u2b80 \u00b1 \u27a6 \u2718 \u26a1 \u2699"
echo ##1##
echo '\ue0b0 \ue0a0 \u2b80 \u00b1 \u27a6 \u2718 \u26a1 \u2699'
echo ##2##
echo -e '\ue0b0 \ue0a0 \u2b80 \u00b1 \u27a6 \u2718 \u26a1 \u2699'
echo ##3##
echo -e "\ue0b0 \ue0a0 \u2b80 \u00b1 \u27a6 \u2718 \u26a1 \u2699"
```


```bash
fc-list : family
```


```bash
fc-match -s emoji
```

```bash
localectl status
```

Outputs:
```bash
   System Locale: LANG=en_US.UTF-8
                  LC_MONETARY=pt_BR.UTF-8
       VC Keymap: us
      X11 Layout: br
       X11 Model: pc104
     X11 Variant: abnt2
     X11 Options: terminate:ctrl_alt_bksp
```


TODO:
https://ostechnix.com/install-nerd-fonts-to-add-glyphs-in-your-code-on-linux/
https://ostechnix.com/find-installed-fonts-commandline-linux/
https://github.com/ryanoasis/nerd-fonts/issues/485#issuecomment-1417572779
https://github.com/ryanoasis/nerd-fonts/issues/485#issuecomment-1417328572


#### pango-view

https://github.com/NixOS/nixpkgs/issues/86601#issuecomment-686243898

#### The echo/print stuff

```bash
echo -e "\U1f3f4\Ue0067\Ue0062\Ue0077\Ue006c\Ue0073\Ue007f"
echo #####
echo -e "\U1f9df\U200d\U2640\Ufe0f"
```

```bash
EMOJIS=(🥯  🦆 🦉 🥓 🦄 🦀 🖕 🍣 🍤 🍥 🍡 🥃 🥞 🤯 🤪 🤬 🤮 🤫 🤭 🧐 🐕 🦖 👾 🐉 🐓 🐋 🐌 🐢)
echo $EMOJIS
```

```bash
UNICORN='\U1F984'; THUMBS_UP='\U1F44D'; echo -e "Riding an ${UNICORN} (${THUMBS_UP})"
```
Refs.:
- TODO

It should output 🍁:
```bash
echo -e '\xF0\x9F\x8D\x81'
```
Refs.:
- https://unix.stackexchange.com/questions/284990/can-a-script-determine-which-characters-a-terminal-can-display



```bash
toon=$'\U1F479'
print -r ${(l:${(m)#toon}:: :)}$'XYZ\n'$toon' ^-- must point to Y'
```
Refs.:
- https://www.zsh.org/mla/users/2020/msg00271.html

```bash
echo "a\uf240 abc"
```
Refs.:
- https://github.com/ryanoasis/nerd-fonts/issues/442#issuecomment-1260423455


```bash
text="Éé"; echo ${#text}

LC_CTYPE=C text="Éé"; echo ${#text}
```
Refs.:
- https://unix.stackexchange.com/questions/245013/get-the-display-width-of-a-string-of-characters/245043#comment1035649_245043
- https://unix.stackexchange.com/questions/245013/get-the-display-width-of-a-string-of-characters/245043#comment1035796_245043

```bash
echo \
'\U1F479' \
'\xF0\x9F\x8D\x81' \
'\U1f9df\U200d' \
'\U1F984' \
'\U1F44D' \
'\U1F9DA' \
'\U1F426' \
'\U1F99C' \
'\U1F996' \
'\U1F420' \
'\U1F41E' \
'\U1F340' \
'\U1F308' \
'\U1F965' \
'\U1F37F' \
'\U1F991' \
'\U1F37A' \
'\U1F692' \
'\U1F6F3' \
'\U26A1' \
'\U1F4A7' \
'\U1F537'
```
Outputs: 👹 🍁 🧟‍ 🦄 👍 🧚 🐦 🦜 🦖 🐠 🐞 🍀 🌈 🥥 🍿 🦑 🍺 🚒 🛳 ⚡ 💧 🔷

Some flags shows:
```bash
echo \
'\U1F3F4\UE0067\UE0062\UE0065\UE006E\UE0067\UE007F' \
'\U1F3F4\UE0067\UE0062\UE0073\UE0063\UE0074\UE007F' \
'\U1f3f4\Ue0067\Ue0062\Ue0077\Ue006c\Ue0073\Ue007f'
```
Outputs: 🏴󠁧󠁢󠁥󠁮󠁧󠁿 🏴󠁧󠁢󠁳󠁣󠁴󠁿 🏴󠁧󠁢󠁷󠁬󠁳󠁿

Some flags are broken:
```bash
echo \
'\U1F3F3\UFE0F\U200D\U1F308' \
'\U1F1E7\U1F1F7'
```

Outputs: 

![image](https://user-images.githubusercontent.com/20312381/222187903-dbbeabca-28f0-4403-8861-8937f520be8c.png)

But copying it directly from terminal to here in browser it renders correctly: 🏳️‍🌈 🇧🇷


The Japan flag, for example:
```bash
echo -e '\U1f1ef\U1f1f5' | hexdump -C
```
Refs.:
- TODO



TODO: test it
```bash
env \
FONTCONFIG_FILE=$PWD/etc-fonts/fonts.conf \
FC_DEBUG=1024 \
pango-view --text="Příliš 😂" --font='"Noto Color Emoji" 20'
```
Refs.:
- https://github.com/NixOS/nixpkgs/issues/86601#issuecomment-686243898



```bash
xterm -fa 'Dank Mono' -fs 11
```
Refs.:
- https://unix.stackexchange.com/a/609609


TODO: impressive `awk`-fu
https://unix.stackexchange.com/a/526681


TODO: it is python code: https://stackoverflow.com/a/37362046

TODO: curl and the '\U0001F514' https://stackoverflow.com/a/55863437

TODO: teste Fira Code + JetBrains Mono
